src/promql/src/functions.rs

```diff
@@ -63,3 +65,123 @@ pub(crate) fn compensated_sum_inc(inc: f64, sum: f64, mut compensation: f64) ->
+            const_y = false;
+        }
+        count += 1.0;
+        let x = time - intercept_time as f64 / 1e3;
+        (sum_x, comp_x) = compensated_sum_inc(x, sum_x, comp_x);
+        (sum_y, comp_y) = compensated_sum_inc(value, sum_y, comp_y);
+        (sum_xy, comp_xy) = compensated_sum_inc(x * value, sum_xy, comp_xy);
+        (sum_x2, comp_x2) = compensated_sum_inc(x * x, sum_x2, comp_x2);
+    }
+
+    if const_y {
+        if !init_y.is_finite() {
+            return (None, None);
+        }
+        return (Some(0.0), Some(init_y));
+    }
+
+    sum_x += comp_x;
+    sum_y += comp_y;
+    sum_xy += comp_xy;
+    sum_x2 += comp_x2;
+
+    let cov_xy = sum_xy - sum_x * sum_y / count;
+    let var_x = sum_x2 - sum_x * sum_x / count;
+
+    let slope = cov_xy / var_x;
+    let intercept = sum_y / count - slope * sum_x / count;
+
+    (Some(slope), Some(intercept))
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn calculate_linear_regression_none() {
+        let ts_array = TimestampMillisecondArray::from_iter(
+            [
+                0i64, 300, 600, 900, 1200, 1500, 1800, 2100, 2400, 2700, 3000,
+            ]
+            .into_iter()
+            .map(Some),
+        );
+        let values_array = Float64Array::from_iter([
+            1.0 / 0.0,
+            1.0 / 0.0,
+            1.0 / 0.0,
+            1.0 / 0.0,
+            1.0 / 0.0,
+            1.0 / 0.0,
+            1.0 / 0.0,
+            1.0 / 0.0,
+            1.0 / 0.0,
+            1.0 / 0.0,
+        ]);
+        let (slope, intercept) = linear_regression(&ts_array, &values_array, ts_array.value(0));
+        assert_eq!(slope, None);
+        assert_eq!(intercept, None);
+    }
+
+    #[test]
+    fn calculate_linear_regression_value_is_const() {
+        let ts_array = TimestampMillisecondArray::from_iter(
+            [
+                0i64, 300, 600, 900, 1200, 1500, 1800, 2100, 2400, 2700, 3000,
+            ]
+            .into_iter()
+            .map(Some),
+        );
+        let values_array =
+            Float64Array::from_iter([10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0]);
+        let (slope, intercept) = linear_regression(&ts_array, &values_array, ts_array.value(0));
+        assert_eq!(slope, Some(0.0));
+        assert_eq!(intercept, Some(10.0));
+    }
+
+    #[test]
+    fn calculate_linear_regression() {
+        let ts_array = TimestampMillisecondArray::from_iter(
+            [
+                0i64, 300, 600, 900, 1200, 1500, 1800, 2100, 2400, 2700, 3000,
+            ]
+            .into_iter()
+            .map(Some),
+        );
+        let values_array = Float64Array::from_iter([
+            0.0, 10.0, 20.0, 30.0, 40.0, 0.0, 10.0, 20.0, 30.0, 40.0, 50.0,
+        ]);
+        let (slope, intercept) = linear_regression(&ts_array, &values_array, ts_array.value(0));
+        assert_eq!(slope, Some(0.010606060606060607));
+        assert_eq!(intercept, Some(6.818181818181818));
+    }
```

Comment:
@waynexia is it guaranteed by caller that input arrays are non-empty? 

Otherwise we need a test case for empty array input to avoid panic

Reply:
Good catch! This guarantee only holds in single-field mode, and may lead to runtime panic when processing a multi-fields table. I filed https://github.com/GreptimeTeam/greptimedb/issues/1378 to track this issue, and I'll re-exam the related parts.